bluetooth/src/androidLibMain/kotlin/scanner/Scanner.kt

```diff
@@ -130,60 +138,95 @@ actual class Scanner internal constructor(
+    override fun stopMonitoringPermissions() {
+        super.stopMonitoringPermissions()
+        monitoringLocationPermissionsJob.get()?.let {
+            if (monitoringLocationPermissionsJob.compareAndSet(it, null)) {
+                it.cancel()
+            }
+        }
     }
 
-    override suspend fun isBluetoothEnabled(): Boolean = bluetoothAdapter?.isEnabled == true // && locationEnabledMonitor.isLocationEnabled()
+    override fun startMonitoringSensors() {
+        super.startMonitoringSensors()
+        if (monitoringLocationEnabledJob.get() != null) return // optimization to skip making a job
+
+        val job = Job(this.coroutineContext[Job])
 
-    override suspend fun requestBluetoothEnable() {
-        // shouldEnableLocation = !locationEnabledMonitor.isLocationEnabled()
-        if (bluetoothAdapter?.isEnabled != true) {
-            bluetoothAdapter?.enable()
-        } else if (shouldEnableLocation) {
-            shouldEnableLocation = false
-            // locationEnabledMonitor.requestLocationEnable()
+        if (monitoringLocationEnabledJob.compareAndSet(null, job)) {
+            locationEnabledMonitor.startMonitoring()
+            launch(job) {
+                locationEnabledMonitor.isEnabled.collect {
+                    checkSensorsEnabledChanged()
+                }
+            }
         }
     }
 
-    private fun checkBluetoothEnabledChanged() =
-        launch(stateRepo.coroutineContext) {
-            val willEnableLocation = shouldEnableLocation
-            shouldEnableLocation = false
-            when {
-                isBluetoothEnabled() -> bluetoothEnabled()
-                willEnableLocation -> {} // locationEnabledMonitor.requestLocationEnable()
-                else -> bluetoothDisabled()
-            }
+    override fun stopMonitoringSensors() {
+        super.stopMonitoringSensors()
+        monitoringLocationEnabledJob.get()?.let {
+            locationEnabledMonitor.stopMonitoring()
+            if (monitoringLocationEnabledJob.compareAndSet(it, null))
+                it.cancel()
         }
+    }
+
+    override suspend fun isPermitted(): Boolean {
+        return super.isPermitted() && locationPermissionRepo.filterOnlyImportant().first() is PermissionState.Allowed
+    }
+
+    override suspend fun areSensorsEnabled(): Boolean = super.areSensorsEnabled() && locationEnabledMonitor.isServiceEnabled
+
+    override fun generateEnableSensorsActions(): List<EnableSensorAction> {
+        if (!isSupported) return emptyList()
+        return listOfNotNull(
+            if (bluetoothAdapter?.isEnabled != true) suspend {
+                bluetoothAdapter?.enable()
+                bluetoothEnabledMonitor!!.isEnabled.first { it }
+            } else null,
+            if (!locationEnabledMonitor.isServiceEnabled) {
+                EnableLocationActivity.showEnableLocationActivity(applicationContext, hashCode().toString())::await
```

Comment:
nitpick: would be better to delegate this to LocationManager, not to duplicate code. The action requestLocationEnable does the same

Reply:
I disagree. LocationManager is a manager used for the LocationState repo. This logic has nothing to do with location because its not interested in the users location. Creating a location manager (and therefore a LocationStateRepo) is overkill.

I've already deferred all validation logic to EnableLocationActivity, so that should suffice